source/extensions/filters/http/ext_proc/mutation_utils.cc

```diff
@@ -76,8 +81,24 @@ absl::Status MutationUtils::applyHeaderMutations(const HeaderMutation& mutation,
     if (!sh.has_header()) {
       continue;
     }
+
+    // Only one of value or value_bytes in the HeaderValue message should be set.
+    if (!sh.header().value().empty() && !sh.header().value_bytes().empty()) {
```

Comment:
Is that possible `sh.has_header ` (i.e. we have key field) but  both `value() and value_bytes()` are empty? It seems so as there is no min_len 1 rule for value fields. 
Is this situation handled by code or can we do early return here?

We can follow up on this in future PRs (if needed)

Reply:
The header value content check is done the logic below: Http::HeaderUtility::headerValueIsValid(header_value)).
From this logic and the HeaderValue proto definition(no min length for value field), it looks to me empty value content is valid, and will pass the follow up mutation rule check by default.

That's the existing logic, and this PR does not change that, i.e., we do not bother if both fields are empty.

Reply:
Yes, I think it will pass the headerValueIsValid check, but do we need to go through the whole check here if it is already empty. This is what I meant by early return.

I understand it is existing logic so I suggested we can follow up on this if needed.

Reply:
An empty header value still needs to go through the mutation rule check. For example, if disallow_all is set to true, that header contains empty header value will still be rejected:

https://github.com/envoyproxy/envoy/blob/f567f66fa68b1533e9daffbdd5c5fd31ee65cba5/source/extensions/filters/common/mutation_rules/mutation_rules.cc#L72

Reply:
But what mutation will happen if sh.header()'s value fields are empty in the first place. That essentially is same as sh.has_header() = false, which skips all the processing here. no?

Reply:
no. sh.has_header() checks whether the header key/value pair exists.

In that pair, the key is more important than value. Like all the system headers check, allow_expression, disallow_expression check are all based on key.

Reply:
I meant what kinds of mutation will happen here in `applyHeaderMutations` if `mutation.set_headers` only has key but not header. 
In other words, what is the mutation outcome not what does check do. Check is the step before mutation. but if there is no outcome at all, does check still important?

Reply:
Yeah, if the ext_proc server sends back a set_header as key/value:   "foo":  "",  this header will be added to the original header list.

Reply:
But this discussion raise a good question that what if the "key" in the header is empty.  

It looks to me an empty header_key will return true in the headerNameIsValie() check in this utility function:
https://github.com/envoyproxy/envoy/blob/3a68249d6b5039f7df5c3f7d740af20a89d97b53/source/common/http/header_utility.cc#L204

Is this expected @yanavlasov ?

Reply:
Hmm, but I am wondering what will such empty value header`"foo": ""` do in practice ?  Probably it is allowed in RFC but don't make too much difference in practice.

Reply:
ext_proc filter is relying on these utilities to make sure the key/value pair are spec compliant.  Adding separate checks in ext_proc filter does not feel right to me.

In either case, let's deal with it separately since it's an existing issue and not related to this adding value_bytes() to support non-utf8 characters.

Reply:
Sure SGTM, let's discuss and address it in next PR (which is also the original intention of my comment)